File: thread_dumper.rb
require 'nokogiri'
require 'open-uri'
require 'fileutils'
require 'parallel'

DUMP_FOLDER = "dump"

FileUtils.mkdir_p DUMP_FOLDER

thread_url = ARGV[0]
page = Nokogiri::HTML(open(thread_url))

subject = page.css('.subject')[1].text
local_folder = "#{DUMP_FOLDER}/#{subject}"
FileUtils.mkdir_p local_folder

images = page.css('.fileText a')
Parallel.each(images, in_processes: 8, progress: "Downloading... #{subject} thread") do |img|
  fnum = img['href'].split('/').last.split('.').first
  local_fname = "#{DUMP_FOLDER}/#{subject}/#{fnum}_#{img.text}"
  remote_url = "http:#{img['href']}"
  unless File.exists? local_fname
    begin
      img_content = open(remote_url).read
    rescue Exception=>e
      puts "Error: #{e}"
    else
      File.open(local_fname, 'wb'){|file| file.write(img_content)}
    end
  end
end

File: thread_dumper.js
const fetch = require('node-fetch')
const cheerio = require('cheerio')
const fs = require('fs')
const term = require('terminal-kit').terminal

const DUMP_FOLDER = "dump"
if (!fs.existsSync(DUMP_FOLDER)) fs.mkdirSync(DUMP_FOLDER)

const thread_url = process.argv[2]

let progress = 0
let progressBar = term.progressBar({
	title: 'dumping:',
	eta: true,
	percent: true
})

fetch(thread_url)
    .then(response => response.text())
    .then(body => {
        let $ = cheerio.load(body)
        
        let subject = $('.subject')[1].children[0].data.replace(/[^\w\s]/gi, '')
        let dir = DUMP_FOLDER + '/' + subject 
        if (!fs.existsSync(dir)) fs.mkdirSync(dir)
        
        let images = $('.fileText a').toArray()
        images.forEach(img => {
            let href = 'http:' + img.attribs['href']
            let file_name = href.split('/').reverse()[0]
            let local_name = file_name.split('.')[0]
                + img.next.data + '.'
                + file_name.split('.')[1]
            fetch(href)
                .then(res => {
                    let dest = fs.createWriteStream(
                        './' + dir + '/' + local_name)
                    res.body.pipe(dest).on('finish', () => {
                        progress += (1 / images.length) + 0.01
                        progressBar.update(progress)
                    })
                })
        })
    })